src/udev/udev-watch.c

```diff
@@ -161,36 +165,462 @@ static int synthesize_change_one(sd_device *dev, sd_device *target) {
+        if (nr < 0)
+                return errno_or_else(EIO);
+
+        const char *whole_devname;
+        r = sd_device_get_devname(dev, &whole_devname);
+        if (r < 0)
+                return r;
+
+        size_t l = strlen(whole_devname);
+        if (l <= 0)
+                return -EINVAL;
+        bool need_p = ascii_isdigit(whole_devname[l - 1]);
+
+        _cleanup_free_ char *node = NULL;
+        if (asprintf(&node, "%s%s%i", whole_devname, need_p ? "p" : "", nr) < 0)
+                return -ENOMEM;
+
+        errno = 0;
+        blkid_loff_t start = sym_blkid_partition_get_start(pp);
+        if (start < 0)
+                return errno_or_else(EIO);
+        assert((uint64_t) start < UINT64_MAX / 512);
+
+        errno = 0;
+        blkid_loff_t size = sym_blkid_partition_get_size(pp);
+        if (size < 0)
+                return errno_or_else(EIO);
+        assert((uint64_t) size < UINT64_MAX / 512);
+
+        /* Use string to support both GPT and DOS. Fortunately, the blkid and kernel uses the same format.
+         * GPT: standard UUID xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+         * DOS: 4 byte disk signature and 1 byte partition xxxxxxxx-xx */
+        _cleanup_free_ char *uuid = NULL;
+        r = strdup_to(&uuid, sym_blkid_partition_get_uuid(pp));
+        if (r < 0)
+                return r;
+
+        _cleanup_(partition_freep) Partition *p = new(Partition, 1);
+        if (!p)
+                return -ENOMEM;
+
+        *p = (Partition) {
+                .node = TAKE_PTR(node),
+                .nr = nr,
+                .start = start,
+                .size = size,
+                .uuid = TAKE_PTR(uuid),
+        };
+
+        r = hashmap_ensure_put(partitions, &partition_hash_ops, p->node, p);
+        if (r < 0)
+                return r;
+
+        TAKE_PTR(p);
+        return 0;
+}
+
+static int enumerate_partitions_by_blkid(sd_device *dev, Hashmap **ret) {
+        int r;
+
+        assert(dev);
+        assert(ret);
+
+        _cleanup_close_ int fd = sd_device_open(dev, O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY);
+        if (fd < 0)
+                return fd;
+
+        if (flock(fd, LOCK_EX|LOCK_NB) < 0)
+                /* The device may be currently locked? If so, let's try again after short delay. */
+                return -errno;
+
+        r = dlopen_libblkid();
         if (r < 0)
-                log_device_debug_errno(dev, r, "Failed to re-read partition table, ignoring: %m");
-        bool part_table_read = r >= 0;
+                return r;
+
+        _cleanup_(blkid_free_probep) blkid_probe b = sym_blkid_new_probe();
+        if (!b)
+                return -ENOMEM;
+
+        errno = 0;
+        r = sym_blkid_probe_set_device(b, fd, /* off= */ 0, /* size= */ 0);
+        if (r != 0)
+                return errno_or_else(ENOMEM);
+
+        sym_blkid_probe_enable_partitions(b, /* enable= */ true);
+        sym_blkid_probe_set_partitions_flags(b, BLKID_PARTS_ENTRY_DETAILS);
```

Comment:
Note that superblock (filesystems, raids, ...) probing is enabled by default. I guess it's wanted as you check for _BLKID_SAFEPROBE_AMBIGUOUS and call blkid_do_safeprobe(). This is what regular udev blkid does. 

If you want to probe only for partition tables, you need to call `blkid_probe_enable_superblocks(b, false)` and, in this case, blkid_do_fullprobe() is good enough.